fluid/conv_seq_to_seq/model.py

```diff
@@ -0,0 +1,238 @@
+                if attention[i] else None)
+            in_channels = out_channels
+
+        self.fc2 = Linear(out_embed_dim)
+        self.fc3 = Linear(dict_size + 1, dropout=dropout)
+
+    def forward(self, prev_output_tokens, prev_positions, encoder_out):
+        '''
+        only works in train mode
+        '''
+        encoder_a, encoder_b = self._split_encoder_out(encoder_out)
+
+        x = self.embed_tokens(prev_output_tokens)
+        x += self.embed_positions(prev_positions)
+        x = Op.dropout(x, self.dropout)
+        target_embedding = x
+
+        # project to size of convolution
+        x = self.fc1(x)
+
+        x = self._transpose_if_training(x)
+
+        # temporal covolutoins
+        avg_attn_scores = None
+        num_attn_layers = len(self.attention)
+        for proj, conv, attention in zip(self.projections, self.convolutions,
+                                         self.attention):
+            residual = x if proj is None else proj(x)
+            x = Op.dropout(x, self.dropout)
+            x = conv(x)
+            x = fluid.nets.glu(x, dim=2)
+
+            if attention is not None:
+                x = self._transpose_if_training(x)
+
+                x, attn_scores = attention(x, target_embedding,
+                                           (encoder_a, encoder_b))
+                attn_scores = attn_scores / num_attn_layers
+                if avg_attn_scores is None:
+                    avg_attn_scores = attn_scores
+                else:
+                    avg_attn_scores += attn_scores
+
+                x = self._transpose_if_training(x)
+
+            x = (x + residual) * math.sqrt(0.5)
+
+        # T x B x C -> B x T x C
+        x = self._transpose_if_training(x)
+
+        # project back to size of vocabulary
+        x = self.fc2(x)
+        x = Op.dropout(x, self.dropout)
+        x = self.fc3(x)
+        # NOTE this is different from fairseq, the predictions are negative and avg_cost will be nan
+        x = Op.softmax(x)
+        return x, avg_attn_scores
+
+    def _transpose_if_training(self, x):
+        x = Op.transpose(x, [0, 1])
+        return x
+
+    def _split_encoder_out(self, encoder_out):
+        encoder_a, encoder_b = encoder_out
+        encoder_a = Op.transpose(encoder_a, [1, 2])
+        result = (encoder_a, encoder_b)
+        return result
+
+
+class AttentionLayer:
+    def __init__(self, conv_channels, embed_dim):
+        self.in_projection = Linear(embed_dim)
+        self.out_projection = Linear(conv_channels)
+
+    def __call__(self, x, target_embedding, encoder_out):
+        return self.forward(x, target_embedding, encoder_out)
+
+    def forward(self, x, target_embedding, encoder_out):
+        '''
+        x is decoder state
+        '''
+        residual = x
+        encoder_a, encoder_b = encoder_out
+        # here just a trick
+        encoder_a = Op.transpose(encoder_a, [1, 2])
+        encoder_b = Op.transpose(encoder_b, [1, 2])
```

Comment:
Is this because `encoder_a, encoder_b = self._split_encoder_out(encoder_out)` convert the shape from `B x T x C` to `B x C x T`. 
We may can remove these two `transpose` and use `x = pd.matmul(x, encoder_a, transpose_y=False)` for dot product calculations.